Mathlib/Tactic/Bound.lean

```diff
@@ -0,0 +1,264 @@
+
+TODO: Kim Morrison noted that we could check for `ℕ` or `ℤ` and try `omega` as well.
+-/
+
+/-- Close numerical goals with `norm_num` -/
+def boundNormNum : Aesop.RuleTac :=
+  Aesop.SingleRuleTac.toRuleTac fun i => do
+    let tac := do Mathlib.Meta.NormNum.elabNormNum .missing .missing .missing
+    let goals ← Lean.Elab.Tactic.run i.goal tac |>.run'
+    if !goals.isEmpty then failure
+    return (#[], none, some .hundred)
+attribute [aesop unsafe 10% tactic (rule_sets := [Bound])] boundNormNum
+
+/-- Close numerical and other goals with `linarith` -/
+def boundLinarith : Aesop.RuleTac :=
+  Aesop.SingleRuleTac.toRuleTac fun i => do
+    Linarith.linarith false [] {} i.goal
+    return (#[], none, some .hundred)
+attribute [aesop unsafe 5% tactic (rule_sets := [Bound])] boundLinarith
+
+/-!
+### `bound` tactic implementation
+-/
+
+/-- Add each provided hypothesis `x` to the context, as by with `have := x` -/
+def addHyps (xs : Array Syntax) : TacticM Unit :=
+  if xs.isEmpty then pure () else Tactic.withMainContext do
+    for x in xs do
+      let v ← elabTerm x none
+      let t ← inferType v
+      liftMetaTactic fun g ↦ do
+        let g ← g.assert `h t v
+        let (_, g) ← g.intro1P
+        return [g]
+
+/-- Aesop configuration for `bound` -/
+def boundConfig : Aesop.Options := {
+  enableSimp := false
+}
+
+end Bound
+
+/-- `bound` tactic for proving inequalities via straightforward recursion on expression structure.
+
+An example use case is
+
+```
+-- Calc example: A weak lower bound for `z ↦ z^2 + c`
+lemma le_sqr_add {c z : ℂ} (cz : abs c ≤ abs z) (z3 : 3 ≤ abs z) :
+    2 * abs z ≤ abs (z^2 + c) := by
+  calc abs (z^2 + c)
+    _ ≥ abs (z^2) - abs c := by bound
+    _ ≥ abs (z^2) - abs z := by bound
+    _ ≥ (abs z - 1) * abs z := by rw [mul_comm, mul_sub_one, ← pow_two, ← abs.map_pow]
+    _ ≥ 2 * abs z := by bound
+```
+
+`bound` is built on top of `aesop`, and uses
+1. Apply lemmas registered via the `@[bound]` attribute
+2. Forward lemmas registered via the `@[bound_forward]` attribute
+3. Local hypotheses from the context
+4. Optionally: additional hypotheses provided as `bound [h₀, h₁]` or similar. These are added to the
+   context as if by `have := hᵢ`.
+
+The functionality of `bound` overlaps with `positivity` and `gcongr`, but can jump back and forth
+between `0 ≤ x` and `x ≤ y`-type inequalities.  For example, `bound` proves
+  `0 ≤ c → b ≤ a → 0 ≤ a * c - b * c`
+by turning the goal into `b * c ≤ a * c`, then using `mul_le_mul_of_nonneg_right`.  `bound` also
+contains lemmas for goals of the form `1 ≤ x, 1 < x, x ≤ 1, x < 1`.  Conversely, `gcongr` can prove
+inequalities for more types of relations, supports all `positivity` functionality, and is likely
+faster since it is more specialized (not built atop `aesop`). -/
+syntax "bound " (" [" term,* "]")? : tactic
+
+elab_rules : tactic
+  | `(tactic| bound $[[$lemmas:term,*]]?) => do
+    let lemmas := (lemmas.getD ⟨#[]⟩).getElems
+    Bound.addHyps lemmas
```

Comment:
If this is really just `have`, then rather than redoing `have` I would make a `macro_rules` that expands this as `have := t1; have := t2; ...; have := tn; bound`.

There's a bug in `addHyps` where if there are any metavariables they get lost, rather than being added as new goals. Using `have` itself would avoid this.

Reply:
That does seem good. Could I hint for what that syntax would look like? I haven’t dealt with macros before.

Reply:
I think this should work:
```
macro_rules
  | `(tactic| bound%$tk [$[$ts],*]) => do
    let haves ← ts.mapM fun (t : Term) => withRef t `(tactic| have := $t)
    `(tactic| ($haves;* ; bound%$tk))
```
(Untested)

Reply:
Works!